[MIPS] Fix error messages when rejecting certain assembly not supported by ISA

jdeguire opened the pull request:
… instructions.

This is a fix I stumbled upon while working on something else. I decided to break it out since it seems like a good "first issue" to submit. I updated the comments in the "wrong error" test files to indicate that the messages are no longer incorrect, but I left the names of the test files alone. I was not sure what to do with those, so I would appreciate thoughts or guidance.

jdeguire committed:
[MIPS] Fix wrong errors being emitted when rejecting certain assembly instructions.

brad0 commented:
@yingopq @wzssyqa

jdeguire commented:
Ping

(Should I add people like "brad0" did above? The automatic comment makes it seem like this is done automatically, so I didn't do that.)

jdeguire commented:
Pong

brad0 commented:
@jdeguire The two people I tagged have been working on MIPS as of lately so tagging so they're aware.

I don't think with how it was tagged it'll get attention with that alone.

@yingopq @wzssyqa

yingopq commented on llvm/lib/Target/Mips/AsmParser/MipsAsmParser.cpp:
```diff
@@ -6410,7 +6410,7 @@ bool MipsAsmParser::parseOperand(OperandVector &Operands, StringRef Mnemonic) {
   // Check if the current operand has a custom associated parser, if so, try to
   // custom parse the operand, or fallback to the general approach.
-  ParseStatus Res = MatchOperandParserImpl(Operands, Mnemonic);
+  ParseStatus Res = MatchOperandParserImpl(Operands, Mnemonic, true);
```
I am not very clear about the purpose of the third parameter ParseForAllFeatures. Some other architectures do not assign a value to this parameter.

jdeguire replied:
As far as I understand, setting the third parameter to True tells the operand parser to accept the operands if they are not valid with the current feature set, but would be valid if the correct feature set were enabled. An example of this would be an instruction that could take floating-point registers if the target device had an FPU. Finding the exact instruction variant that takes those operands is done after parsing them and it is at that point the "instruction requires a CPU feature not currently enabled" error is triggered. The default for that third parameter is False, which in this case causes the operand parser to give up if an operand is not valid in the current feature set. The symptom is that you will receive a less useful "invalid operand" error instead of one telling you that you do not have the correct features enabled.

jdeguire replied:
I added a comment here to explain what the third parameter does. Hopefully that is helpful.

yingopq commented:
Sorry for replying so late because I'm not very familiar with this area.

jdeguire commented:
No problem, I appreciate your response!

jdeguire commented:
Ping!

Apologies for the late check-in. Does my explanation of what the change is doing help? I could try adding a comment to the code if that would be better.

yingopq commented:
Can we give a more detailed error, such as what feature is not set?

MaskRay commented on llvm/test/MC/Mips/eva/invalid-noeva-wrong-error.s:
```diff
@@ -19,51 +19,51 @@
-        swre      $s4,255($13)         # CHECK: :[[@LINE]]:{{[0-9]+}}: error: invalid operand for instruction
-        swre      $s4,-256($13)        # CHECK: :[[@LINE]]:{{[0-9]+}}: error: invalid operand for instruction
-        swre      $s2,86($14)          # CHECK: :[[@LINE]]:{{[0-9]+}}: error: invalid operand for instruction
+        cachee    31, 255($7)          # CHECK: :[[@LINE]]:{{[0-9]+}}: error: instruction requires a CPU feature not currently enabled
```
While updating messages, switch to `[[#@LINE]]`. `[[@LINE]]` is deprecated FileCheck syntax. Prefer `[[#]]` to `{{[0-9]+}}`

jdeguire replied:
I updated the messages I modified in the the most recent commit. Should I update the other messages in these files? I can do that here or as another pull request later.

jdeguire committed:
Update based on comments from LLVM PR 94695.

-- Add a comment to MipsAsmParser explaining what the third parameter in the call to 'MatchOperandParseImplr()' does.
-- Update the error message checks to use newer syntax. This updates only the messages I had originally modified.

jdeguire commented:
> Can we give a more detailed error, such as what feature is not set?

It looks like some other targets do this, so I should be able to add that to MIPS, too. Should something like that be a part of this PR or a later one?

jdeguire committed:
Fix formatting issue flagged by clang-format.

jdeguire commented:
Ping.

yingopq commented:
LGTM.

wzssyqa commented:
LGTM.
